source/views/building-hours/detail/index.js

```diff
@@ -11,17 +11,45 @@ import moment from 'moment-timezone'
 import {BuildingDetail} from './building'
 import {CENTRAL_TZ} from '../lib'
 import type {TopLevelViewPropsType} from '../../types'
+import {FavoriteButton} from '../../components/nav-buttons'
+import {connect} from 'react-redux'
+import {type ReduxState} from '../../../flux'
 
-type Props = TopLevelViewPropsType & {
-  navigation: {state: {params: {building: BuildingType}}},
+type ReduxStateProps = {
+  favoriteBuildings: Array<BuildingType>,
 }
 
+type Props = TopLevelViewPropsType &
+  ReduxStateProps & {
+    navigation: {
+      state: {
+        params: {building: BuildingType, favoriteBuildings: BuildingType[]},
+      },
+    },
+  }
+
 type State = {intervalId: number, now: moment}
 
 export class BuildingHoursDetailView extends React.PureComponent<Props, State> {
   static navigationOptions = ({navigation}) => {
+    const building = navigation.state.params.building
+    let favoriteBuildings = navigation.state.params.favoriteBuildings
     return {
-      title: navigation.state.params.building.name,
+      title: building.name,
+      headerRight: (
+        <FavoriteButton
+          favorited={favoriteBuildings.includes(building)}
+          navigation={navigation}
+          onFavorite={() => {
```

Comment:
I'd like to pull this logic out of here, probably into a function in parts/building-hours.js? Said function would be called as part of the saveBuildingFavorites function.

Reply:
Essentially, then, we avoid mutating the array here entirely and call onSaveBuildingFavorites or whatever it's called (sorry, reviewing from my phone atm) instead. Redux will cause a re-render or the screen when the list of favorited buildings changes, so we'll know that the building is favorited because it suddenly will be (if that makes sense; I can try to explain better if you want me to.)

Reply:
You'll also need to add the dispatch call / map thing in here, instead of in StatefulList, because that's how this component will have access to the function.